Implementation of ipython widget

chmp opened the pull request:
This is an update of the original PR (#94). After fighting a bit with webpack, I think I found a solution to use optional dependencies. For the weak dependencies on iypwidgets, the solution was to

- add the webpack types to typescript
- use [`resolveWeak`](https://webpack.js.org/api/module-methods/#requireresolveweak) to query at module definition time whether the ipywidgets javascript module is available 
- only if so, require the module and define the widget implementation

The [added example notebook](https://github.com/chmp/ipyvega/blob/feature/widget/notebooks/Widget.ipynb) demonstrates:

- a standard vega plot
- a vega widget with live updates
- a vega widget embedded in a larger ipywidgets layout with additional interactivity

I verified for the production build (`yarn build`) that the vega plot works with and without ipywidgets installed. The later two parts give an import error when ipywidgets is not installed.

At the moment I am unhappy that Javascript errors are not translated back into python errors. For example, specifying the wrong name in update fails silently at the moment. My goal would be to have some kind of a lightweight exception translation mechanism to surface these errors to the user directly. 

---

For reference, the example of the vega widget with interactivity looks like:

![Example Vega Widget in Layout](https://user-images.githubusercontent.com/949987/57017758-d207f180-6c20-11e9-8254-d8a72639eed7.png)

---

The original PR was:

> As discussed in #93, this PR contains an implementation for an IPython Vega widget. In addition to the existing display system, it allows
> 
>     to put altair charts in a larger UI with other widgets,
>     to change the chart without re-displaying the widget, and
>     to dynamically update the data via changesets.
> 
> I struggled a bit with webpack to get it to only optionally define the widget component. Not sure whether my solution is idiomatic.

chmp commented:
Oh and I rebased to master.

domoritz commented:
This is a great start. Let me know when you want me to review. Will this allow us to see the version of Vega and Vega-Lite from Python?

Also, what will it take to implement this in JupyterLab? It would be great to have feature parity there.

domoritz commented:
cc @saba9

chmp commented:
> This is a great start. Let me know when you want me to review. 

Thanks for the feedback. I wanted to open a PR early so you can track what's roughly going on. There are some things to be ironed out (the error translation mechanism, a rendering bug, a race condition where immediate updates are ignored, jupyter lab, ...), before this implementation has a state that is mergeable. While I would enjoy early feedback, I fear the code clould still change a bit. 

> Will this allow us to see the version of Vega and Vega-Lite from Python?

In prinicple yes. Allthough the calls would be async and without switching out the event loop waiting for the result seems to deadlock the interface. At least this is what my initial attempts at error translation resulted in. 

Could you maybe elaborate a bit more, what the goal would be? Since vega seems to be bundled with the plugin code, you should be able to determine the vega version at package build time, shouldn't you?  

> Also, what will it take to implement this in JupyterLab? It would be great to have feature parity there.

I will have a look. The initial test was not very successful, or at all :) . I'll give it another go after reading the docs more thoroughly, but most probably will open a ticket in jupyterlab to clarify what the correct approach would be. This whole javascript bundling thing is driving me insane.

domoritz commented:
> Could you maybe elaborate a bit more, what the goal would be?

I think it's just a small test case. The real advantage is if we can access the selections from the python kernel.

> This whole javascript bundling thing is driving me insane.

It's super annoying and also keeps changing. At some point, you get the hang of most of the quirks.

chmp commented:
Hm. Re. jupyterlab: can it be that alreay the current version does not work with jupyterlab? If I do a ``Vega(spec)``, I get ``Javascript Error: require is not defined``. I used the PyPI package.

> I think it's just a small test case. The real advantage is if we can access the selections from the python kernel.

Not sure how the vega API looks for this feature. As long as it is asynchronous, you can pass events between python and javascript. For example to listen to the click event, you can do the following:

```javascript
this.view.addEventListener('click', (event: any, item: any) => {
  this.send({"type": "click"}) 
});
```

```python
@widget.on_msg
def _(_1, ev, _2):
    if ev["type"] == "click":
        ...
```

chmp commented:
Okay this was one long fall into the rabbit hole. Following the discussion on [this issue](https://github.com/bokeh/bokeh/issues/6700), [this issue](https://github.com/jupyter-widgets/ipywidgets/issues/1799) and [this PR](https://github.com/jupyter/notebook/pull/1866), my current understanding how to make ipyvega work across notebook and lab is the following: 

- inside jupyter notebook, registers a custom mimetype handler for the vega/vega-lite mimtetypes
- inside jupyter lab, use the render logic shipped with jupyerlab for the vega/vega-lite mimetypes
- inside jupyter notebook, register the widget code as is done now 
- inside jupyter lab, use a extension, published to npm, that contains the widget code
- in python, include the custom mime-type in the mime-bundle of `Vega` and `VegaLite` 
- in python, **do not** include the javascript code in the mime-bundle

<details>
A simple implementation of the mimebundle could look like:

```python
class VegaLite:
    ...
    def _repr_mimebundle_(self, include=None, exclude=None):
        return {'application/vnd.vegalite.v1+json': self.spec}
```

A simple implementation of the mimetype renderer could look like:

```javascript
var MIME_TYPE = 'application/vnd.vegalite.v1+json';

require(["nbextensions/jupyter-vega/index"], function(vega) {    
    function render_vega(spec, md, element) {
        var toinsert = this.create_output_subarea(md, 'output_json rendered_html', MIME_TYPE);
        element.append(toinsert);
        
        var el = toinsert[0];
        
        vega.vegaEmbed(el, spec, {
            defaultStyle: true,
            loader: { http: { credentials: "same-origin" } },
            mode: 'vega-lite',
          })
        
        return toinsert;
    }
    
    OutputArea.prototype.register_mime_type(MIME_TYPE, render_vega, true);
});
```
</details>

domoritz commented:
> Re. jupyterlab: can it be that alreay the current version does not work with jupyterlab?

Jupyterlab does not use the `vega` package. It has its own implementation in https://github.com/jupyterlab/jupyterlab/tree/master/packages/vega5-extension

chmp commented:
I update the code to reflect updates correctly that are sent before the widget is displayed. As described in the source code:

```python
    # Implementation note: there is a small delay between defining the widget
    # and its display in the frontend. Any message sent during this time
    # interval will be silently ignored by the client. To ensure all updates
    # are handled, they are buffered on the python side until the widget is
    # first displayed. The buffer is the `_pending_updates` attribute and the
    # display state is reflected by the `_displayed` attribute.
```

The only major thing is missing is some error message display (e.g., missing dataset, invalid filter expression). I will add an warning box directly to the widget to communicate such errors to the user.

> Jupyterlab does not use the vega package. It has its own implementation in https://github.com/jupyterlab/jupyterlab/tree/master/packages/vega5-extension

Yes. I only feel it is unfortunate, that if user tries to use ipyvega from jupyterlab it results in an error. In the current state, the readme's "you don't need to install ..." is somewhat misleading. It is rather "you must not use ...". The  issues I referenced above, discuss a way that would make ipyvega work in both environments, transparently falling back on the built-in support in jupyterlab. But I guess this discussion is orthogonal to this issue / PR. If I have a bit more time, I will open a separate issue with a more detailed reasoning. In any case, I don't have strong feelings about this issue.

chmp commented:
I added a simple mechanism to communicate javascript errors to the user. It looks like that:

![Screenshot of error handling](https://user-images.githubusercontent.com/949987/57327974-73ef7880-7110-11e9-9021-2feb83adf5ca.png)

From my side this would be, what I had in mind to implement. I will rebase to master and would welcome your feedback afterwards :).

chmp commented:
Rebased to master.

domoritz commented on src/index.ts:
```diff
@@ -92,3 +92,97 @@ export function render(
+      this.el.appendChild(this.errorElement);
+
+      const reembed = () => {
+        this.view = null;
```
```suggestion
	    if (this.view) {
	    	view.finalize();
	    }
        this.view = null;
```

domoritz commented on vega/widget.py:
```diff
@@ -0,0 +1,122 @@
+    _spec_source = Unicode('null').tag(sync=True)
+
+    @classmethod
+    def from_chart(cls, chart):
```
This seems like a hack for Altair, no? Wouldn't it be better to instead ask people to write

```
widget = VegaWidget(chart.to_dict())
widget.update(remove='datum.t < 5', insert=[{...}, {...}])
```

cc @jakevdp

chmp replied:
You're right. I'll update the doc string.

domoritz commented on src/index.ts:
```diff
@@ -92,3 +92,97 @@ export function render(
+  if (ev.type != "update") {
+    return null;
+  }
+  // TODO: perform more checks and validate the event
```
What kind of checks and for what? I don't really know how to work with this TODO.

chmp replied:
Godd point. I also needed a minute to refamilarize myself with my thought process :).

Essentially, the idea was validate that the event really conforms to the `WidgetUpdateMessage` type, i.e., also checking the `updates` array. With a bit of distance, I'm no longer sure, whether it really is relevant, as the API is purely used internally.

domoritz replied:
Please update the TODO accordingly then.

chmp commented:
@domoritz I finally got time to work on this again.

- I fixed some a small error in this view.finalize call and rebuild the static js
- I removed the `from_chart` constructor
- I rephrased the TODO marker on checkWidgetUpdate

domoritz commented on src/index.ts:
```diff
@@ -92,3 +92,102 @@ export function render(
+          return;
+        }
+
+        vegaEmbed(this.viewElement, spec)
```
Add support for EmbedOptions as above.

domoritz replied:
We should probably also set the loader as above.

You can ignore the mode since you are setting it via the spec. However, I'd say that the docs need to be updated to explain how to switch from VL to Vega.

chmp replied:
I added the EmbedOptions in the same way as the spec is handled, i.e., they can be set at construction and dynamically. I also added a corresponding section to the example notebook (and discovered / fixed a small bug with dynamic spec updates).

Regarding the mode. TBH, I was kind of unaware of the mode option, just copied the spec from the existing example notebooks and was happy it worked. 

If you like, I can add the same split into `VegaWidget` and `VegaLiteWidget` as for the non-widget code parts. Via subclassing the changes should be minimal.

domoritz replied:
Nah, I think relying on $schema is enough, I think.

chmp replied:
I added a sentence about the ``$schema`` property to the docstring.

domoritz commented:
Please rebase to master so I can merge.

chmp commented:
I rebased to master and added a sentence about the ``$schema`` property.

domoritz commented:
🎉 Thank you so much for your contribution.

chmp commented:
Thanks a lot for merging :).

domoritz commented:
Unfortunately, this broke a bunch of users: https://github.com/vega/ipyvega/issues/105
